### PR TITLE
Llk refactor uplift gs

### PR DESCRIPTION
**Pull request by rtawfik01**

### CODE REVIEW

**Review by muthutt — COMMENTED** [review]
I'm trying to run some tests on GH and WH manually and inspect your code; should have review by end of day and you should be open to submit on Tuesday this week (tomorrow)

**Review by muthutt — APPROVED** [review]
LGTM - to unblock large PR - but request to address minor issues

File: tt_metal/hw/ckernels/wormhole_b0/common/inc/ckernel.h
Comment (rtawfik01) [reply tt_metal/hw/ckernels/wormhole_b0/common/inc/ckernel.h:39]: Hi @muthutt , its because the rest of the defines all use `#ifndef`, so changing the `#ifdef` to `#if defined` provides a more visible difference than using `ifdef` and `ifndef` sequentially in the same file (just easier for debugging). If its an issue though for the coding standard I can definitely change it.

File: tt_metal/hw/ckernels/grayskull/common/inc/ckernel_globals.h
Comment (mywoodstock) [inline tt_metal/hw/ckernels/grayskull/common/inc/ckernel_globals.h:15]: should this be a pointer?

File: tt_metal/hw/ckernels/grayskull/llk_lib/llk_math_eltwise_binary.h
Comment (DrJessop) [inline tt_metal/hw/ckernels/grayskull/llk_lib/llk_math_eltwise_binary.h:54]: please remove comments.

File: tt_metal/hw/ckernels/grayskull/llk_lib/llk_pack_common.h
Comment (DrJessop) [inline tt_metal/hw/ckernels/grayskull/llk_lib/llk_pack_common.h:26]: Please remove all perf dump code.

File: tt_metal/hw/ckernels/grayskull/llk_lib/llk_pack_common.h
Comment (DrJessop) [inline tt_metal/hw/ckernels/grayskull/llk_lib/llk_pack_common.h:212]: Not following why this is templatized on untilize. If it's just for the sake of bringing this up to date with latest buda backend, that's fine I guess.

File: tt_metal/hw/ckernels/grayskull/llk_lib/llk_unpack_AB.h
Comment (DrJessop) [inline tt_metal/hw/ckernels/grayskull/llk_lib/llk_unpack_AB.h:141]: Will mention one more time, but this should apply everywhere. There should not be any perf dump code blocks, we don't support the flow used in budabackend.

File: tt_metal/hw/ckernels/grayskull/llk_lib/llk_pack_untilize.h
Comment (mywoodstock) [inline tt_metal/hw/ckernels/grayskull/llk_lib/llk_pack_untilize.h:98]: these have been removed?

Comment (mywoodstock) [reply tt_metal/hw/ckernels/grayskull/llk_lib/llk_pack_untilize.h:98]: ok nvm, i see its moved :)

**Review by DrJessop — APPROVED** [review]
Approving under the following condition:
- Remove any instances of perf dump code
- Remove commented out code

I didn't spend time reviewing code quality since this just seems like a straight up uplift to have up to date budabackend code, after which we'll be commonizing the two stacks. Let me know if you want a more thorough review if that wasn't the only scope of this PR.

**Review by mywoodstock — APPROVED** [review]
lgtm!

File: tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/llk_math_eltwise_unary_sfpu_negative.h
Comment (rtawfik01) [reply tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/llk_math_eltwise_unary_sfpu_negative.h]: Fixed, first iteration was removed in the other PR for wormhole_b0

File: tt_metal/hw/ckernels/wormhole_b0/common/inc/ckernel.h
Comment (rtawfik01) [reply tt_metal/hw/ckernels/wormhole_b0/common/inc/ckernel.h]: Hey @muthutt , since the above file is common between buda & metal and will be in its own repo soon, we will keep that define to keep it parity until the common submodule

File: tt_metal/hw/ckernels/grayskull/llk_lib/llk_pack_common.h
Comment (rtawfik01) [reply tt_metal/hw/ckernels/grayskull/llk_lib/llk_pack_common.h:212]: yes this is common with buda, but if its not used we will clean it up there too

File: tt_metal/hw/ckernels/grayskull/llk_lib/llk_pack_untilize.h
Comment (rtawfik01) [reply tt_metal/hw/ckernels/grayskull/llk_lib/llk_pack_untilize.h:98]: yes they are in llk_pack_api.h now

File: tt_metal/hw/ckernels/grayskull/common/inc/ckernel_globals.h
Comment (rtawfik01) [reply tt_metal/hw/ckernels/grayskull/common/inc/ckernel_globals.h:15]: This l1 buffer is used by the reduce op and I believe it meant to be static l1 memory, but Ill add @acejkov to confirm.

File: tt_metal/hw/ckernels/grayskull/llk_lib/llk_unpack_AB.h
Comment (rtawfik01) [reply tt_metal/hw/ckernels/grayskull/llk_lib/llk_unpack_AB.h:141]: agree we will have to remove perf dump code, but problem is right now is that buda perf dump hooks are ingrained in the llk_lib kernels, and will cause all buda perf test sweeps to fail if removed now (because this file will be parity with buda). But I agree @acejkov and I will review and separate these perf hooks to be in the buda llk_api files once these files are commonalized and we can make sure buda regressions don't fail.

File: tt_metal/hw/ckernels/grayskull/llk_lib/llk_pack_common.h
Comment (rtawfik01) [reply tt_metal/hw/ckernels/grayskull/llk_lib/llk_pack_common.h:26]: answered in the other perf dump comment

File: tt_metal/hw/ckernels/grayskull/llk_lib/llk_unpack_AB.h
Comment (acejkov) [reply tt_metal/hw/ckernels/grayskull/llk_lib/llk_unpack_AB.h:141]: there might be an option to reuse perf dump tools for metal. we keep as integral part of the LLK and higher level stack can decide to use it or not

File: tt_metal/hw/ckernels/grayskull/common/inc/ckernel_globals.h
Comment (acejkov) [reply tt_metal/hw/ckernels/grayskull/common/inc/ckernel_globals.h:15]: yes, we need it for buda

File: tt_metal/hw/ckernels/grayskull/llk_lib/llk_math_eltwise_binary.h
Comment (rtawfik01) [reply tt_metal/hw/ckernels/grayskull/llk_lib/llk_math_eltwise_binary.h]: Done

File: tt_metal/hw/ckernels/grayskull/common/inc/ckernel.h
Comment (davorchap) [inline tt_metal/hw/ckernels/grayskull/common/inc/ckernel.h:335]: Are these use in the new set of GS LLKs? Or are these for debug infra?

File: tt_metal/hw/ckernels/grayskull/common/inc/ckernel.h
Comment (davorchap) [inline tt_metal/hw/ckernels/grayskull/common/inc/ckernel.h:444]: The PERF_DUMP infra -- this is BBE specific?
We can have it for now in `ckernel.h` , but perhaps it can be factored out to its own file later on, so that if not used we don't have to compile it.

File: tt_metal/hw/ckernels/grayskull/common/inc/ckernel_addr_map.h
Comment (davorchap) [inline tt_metal/hw/ckernels/grayskull/common/inc/ckernel_addr_map.h:6]: If also BBE specific, we should plan how to factor out to a separate file.

File: tt_metal/hw/ckernels/grayskull/common/inc/ckernel.h
Comment (rtawfik01) [reply tt_metal/hw/ckernels/grayskull/common/inc/ckernel.h:335]: These are for debug infra

File: tt_metal/hw/ckernels/grayskull/llk_lib/llk_unpack_AB_matmul.h
Comment (davorchap) [inline tt_metal/hw/ckernels/grayskull/llk_lib/llk_unpack_AB_matmul.h:83]: I believe hoisting this outside the function and before the matmul rt_dim/ct_dim loops did improve performance. 
@mywoodstock measured BERT-L on this branch at 411 sent/sec, so perhaps it's getting auto-hoisted in the refactored code.

File: tt_metal/hw/ckernels/grayskull/llk_lib/llk_unpack_AB_matmul.h
Comment (davorchap) [inline tt_metal/hw/ckernels/grayskull/llk_lib/llk_unpack_AB_matmul.h:233]: hoisting of the `cfg` ptr outside of the loops in the old implementation

File: tt_metal/hw/ckernels/grayskull/common/inc/ckernel.h
Comment (rtawfik01) [reply tt_metal/hw/ckernels/grayskull/common/inc/ckernel.h:444]: PERF_DUMP is currently used by BBE, but the perf infra in the llks can be re-used by either stacks if required since it is general

File: tt_metal/hw/ckernels/grayskull/common/inc/ckernel_pcbuf.h
Comment (rtawfik01) [reply tt_metal/hw/ckernels/grayskull/common/inc/ckernel_pcbuf.h:13]: Yes it is in the firmware

**Comment by davorchap** [comment]
overall it looks great, thank you!

* made comments about some code that is not used in Metal, so perhaps could be refactor into separate files so that compile can be skipped
* llk_matmul_tile in the block mode (rt_dim, ct_dim) has changed, ie `cfg` ptr is not manually hoisted out of the rt/ct loops, but there doesn't seem to be a perf drop based on Abhinav's measurements. I'm not sure `cfg` can  be auto-hoisted given that it is volatile.


@mywoodstock measurements: 
ResNet50-b20 
"AVG DEVICE KERNEL SAMPLES/S": 6156.77346282875,
    "MIN DEVICE KERNEL SAMPLES/S": 6152.297667848523,
    "MAX DEVICE KERNEL SAMPLES/S": 6164.067125458182,
 
BERT-L, b12
 "AVG DEVICE KERNEL SAMPLES/S": 411.0532633569959,
    "MIN DEVICE KERNEL SAMPLES/S": 410.89907045042617,
    "MAX DEVICE KERNEL SAMPLES/S": 411.16637415605527,

**Review by davorchap — APPROVED** [review]

File: tt_metal/hw/ckernels/grayskull/llk_lib/llk_unpack_AB_matmul.h
Comment (rtawfik01) [reply tt_metal/hw/ckernels/grayskull/llk_lib/llk_unpack_AB_matmul.h:83]: I think when we commonlize the code, we can try taking the cfg pointer specifically out and see if it helps performance for metal/buda, but I think the bigger impact was the address calculations being in the higher level loops:
```
    const std::uint32_t operandA_id = get_operand_id(operandA);
    const std::uint32_t operandB_id = get_operand_id(operandB);

    std::uint32_t base_address_a = cb_interface[operandA_id].fifo_rd_ptr - 1;
    std::uint32_t base_address_b = cb_interface[operandB_id].fifo_rd_ptr - 1;
    std::uint32_t unpA_src_format = (std::uint32_t)unpack_src_format[operandA_id];
    std::uint32_t unpB_src_format = (std::uint32_t)unpack_src_format[operandB_id];

    for (std::uint32_t rt=0; rt<rt_dim; rt++) {
        std::uint32_t offset_address_a = MUL_TILE_SIZE_AND_INDEX<true>(unpA_src_format, (tile_index_a + rt*kt_dim));
        std::uint32_t address_a = base_address_a + offset_address_a;

        for (std::uint32_t ct=0; ct<ct_dim; ct++) {

            std::uint32_t offset_address_b = MUL_TILE_SIZE_AND_INDEX<true>(unpB_src_format, (tile_index_b+ct));
            std::uint32_t address_b = base_address_b + offset_address_b;

            _llk_unpack_AB_matmul_(
                address_a,
                address_b
            );
```

because we did see before that being able to push those address calculations to replay buffers/outerloops in wormhole gave a lot of performance gains. 

But currently because llk_unpack_AB_matmul.h file is common for both buda/metal, and buda is still calculating by tile, then the cfg pointer will have to stay in the per tile calculation until we commonlize/back-propagate back to buda

**Pull request merged**
